Add extra envs to Helmchart values.yaml

mocdaniel opened the pull request:
This PR fixes #287 by introducing a new Helmchart value called `extraEnvs` to the `manager` container of the operator's deployment.

It allows users to define arbitrary environment variables they might need inside their operator, e.g. `HTTP_PROXY`/`NO_PROXY` settings.

I saw no use in adding the same configuration to the `kube-rbac-proxy` container, let me know if I should adapt this container's configuration, too.

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/vault-secrets-operator?pullRequest=304)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/vault-secrets-operator?pullRequest=304)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)

<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/vault-secrets-operator?pullRequest=304).</sub>

benashz commented on chart/templates/deployment.yaml:
```diff
@@ -83,6 +83,10 @@ spec:
           value: {{ .Values.controller.kubernetesClusterDomain }}
+        {{- range .Values.controller.manager.extraEnvs }}
+        - name: {{ .name }}
+          value: {{ .value | quote }}
```
Could this result in double quoting the value e.g.: `.value = "foo"`

mocdaniel replied:
No, `quote` doesn't double quote already quoted strings. It's best practice to always pipe inputs expected to be strings or env values to `quote`, [according to the Helm documentation](https://helm.sh/docs/howto/charts_tips_and_tricks/#quote-strings-dont-quote-integers).

benashz reviewed (commented):
Thanks for the PR! I've added a few comments/suggestions.  Also, it would be great if you could extend the `bats` tests here: https://github.com/hashicorp/vault-secrets-operator/tree/main/test/unit

mocdaniel commented:
> Thanks for the PR! I've added a few comments/suggestions. Also, it would be great if you could extend the `bats` tests here: https://github.com/hashicorp/vault-secrets-operator/tree/main/test/unit

I added four new tests, one for the Chart's default configuration (no extraEnv), and one for each of the allowed env types (string, number, quoted string).

benashz commented on test/unit/deployment.bats:
```diff
@@ -352,3 +352,67 @@ load _helpers
+    [ "${actual}" = 'QUOTED_ENV' ]
+    actual=$(echo "$object" | yq '.[3].value' | tee /dev/stderr)
+    [ "${actual}" = 'noquotesneeded' ]
+}
```
Looks like we lost the end of file newline.

benashz commented on test/unit/deployment.bats:
```diff
@@ -352,3 +352,67 @@ load _helpers
+      yq '.spec.template.spec.containers[1].env | select(documentIndex == 1)' |  \
+      tee /dev/stderr)
+
+    local actual=$(echo "$object" | yq '.[3].name' | tee /dev/stderr)
```
We should probably add a length check here to make sure we are not picking up any unexpected values.

mocdaniel replied:
Could you elaborate on this? I don't get it, I'm afraid. I am already checking default behaviour (no `extraVar` set) in a dedicated check, am I missing something?

mocdaniel commented on test/unit/deployment.bats:
```diff
@@ -352,3 +352,67 @@ load _helpers
+      yq '.spec.template.spec.containers[1].env | select(documentIndex == 1)' |  \
+      tee /dev/stderr)
+
+    local actual=$(echo "$object" | yq '.[3].name' | tee /dev/stderr)
```
I will add this.

mocdaniel commented on test/unit/deployment.bats:
```diff
@@ -352,3 +352,67 @@ load _helpers
+    local object=$(helm template  \
+      -s templates/deployment.yaml  \
+      --set 'controller.manager.extraEnv[0].name=HTTP_PROXY'  \
+      --set 'controller.manager.extraEnv[0].value=http://proxy.example.com/'  \
```
I will add this, too.

mocdaniel commented:
Regarding the quoting issue:

In your example, you *explicitly* quote the value (the equivalent to `value: \"42\"` in YAML if I'm not mistaken). Those quotes get preserved, as they are interpreted as intentional, since YAML doesn't require quoting of strings.

My test cases covers a *normal* quoted string in a `values.yaml`, as some users like to quote all their strings (the equivalent to `value: "some"`), as well as unquoted strings (`value: some`), and integers (`value: 42`).

However, I looked up some popular Helmcharts and how they handle envs, and the majority in fact doesn't quote the envs. So I would leave the final decision whether to quote envs or not to you.

Once this question is settled I will refactor the changes to the chart and tests to reflect our final decision.

benashz commented:
> Regarding the quoting issue:
> 
> In your example, you _explicitly_ quote the value (the equivalent to `value: \"42\"` in YAML if I'm not mistaken). Those quotes get preserved, as they are interpreted as intentional, since YAML doesn't require quoting of strings.

Indeed. Imagine the case where the value contains whitespace, it would need to be quoted on for the shell to properly handle it.

> 
> My test cases covers a _normal_ quoted string in a `values.yaml`, as some users like to quote all their strings (the equivalent to `value: "some"`), as well as unquoted strings (`value: some`), and integers (`value: 42`).
> 

I think it is best to treat all values as strings.

> However, I looked up some popular Helmcharts and how they handle envs, and the majority in fact doesn't quote the envs. So I would leave the final decision whether to quote envs or not to you
> 
> Once this question is settled I will refactor the changes to the chart and tests to reflect our final decision

Let's drop the  ` | quote`. Thanks!

mocdaniel commented:
I dropped `quote` and adjusted the tests. All comments should be resolved now.

benashz reviewed (approved):
LGTM!

Thank you for your contribution to HashiCorp!
